### PR TITLE
Enso Date shall be converted to java.time.LocalDate when passed to Java

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

This PR provides a unit test, an integration test and code changes to guarantee automatic conversion of `java.time.LocalDate` to Enso `Date` atom and back.

Originally (after the first few commits) the following test failed:
```bash
sbt:runtime> testOnly *DateTest*
 - should compare dates in java *** FAILED ***
[info]     org.enso.interpreter.test.InterpreterException: class com.oracle.truffle.polyglot.PolyglotMap cannot be cast to class java.time.chrono.ChronoLocalDate (com.oracle.truffle.polyglot.PolyglotMap is in module org.graalvm.truffle of loader 'app'; java.time.chrono.ChronoLocalDate is in module java.base of loader 'bootstrap')
[info]     at java.base/java.time.LocalDate.compareTo(LocalDate.java:139)
[info]     at <enso>.argument<1>(<interactive_source>:3)
[info]     at <enso>.<eval>(Unknown)
[info]     at <enso>.Debug.eval(Unknown)
[info]     at <enso>.Test.main(Test:6)
[info]     at org.graalvm.sdk/org.graalvm.polyglot.Value.execute(Value.java:841)
```

then this PR seeks to provide a fix. The PR has been [enhanced with another test](https://github.com/enso-org/enso/pull/3374/commits/d4a8e31fc856be56095a62dcf9b117c5fbd09aa1) that sends `java.time.LocalDate` back to Enso and fails with:

```bash
 - should check java date has enso methods *** FAILED ***
[info]     org.enso.interpreter.test.InterpreterException: Method `year` of java.time.LocalDate could not be found.
[info]     at <enso>.Test.main(Test:7)
[info]     at org.graalvm.sdk/org.graalvm.polyglot.Value.execute(Value.java:841)
```

Now all the unit as well as integration tests pass OK.

### Checklist

Please include the following checklist in your PR:

- [ x ] The documentation has been updated if necessary.
- [ x ] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- All code has been tested:
  - [ x ] Unit tests have been written where possible.

### CODE REVIEW

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]: Delegating to `fields[0]` is a bit wild - I guess I should first learn what's the meaning of `fields` array.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]: Right now I am putting this call behind `TruffleBoundary` and using `InteropLibrary.getUncached()` - this is not going to be fast - it is possible to speed things up, if necessary.

File: engine/runtime/src/test/scala/org/enso/interpreter/test/semantic/DateTest.scala
Comment (JaroslavTulach) [inline engine/runtime/src/test/scala/org/enso/interpreter/test/semantic/DateTest.scala]: I had to find a method that takes `LocalDate` as parameter and pass `ensodate` there - that's the problematic use-case. I thought `LocalDate.compareTo` would do, but it doesn't. The [compareTo](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/time/LocalDate.html#compareTo(java.time.chrono.ChronoLocalDate)) takes `ChronoLocalDate` - an interface that `LocalDate` implements. That confuses Truffle interop...

File: engine/runtime/src/test/scala/org/enso/interpreter/test/semantic/DateTest.scala
Comment (JaroslavTulach) [inline engine/runtime/src/test/scala/org/enso/interpreter/test/semantic/DateTest.scala]: ...at the end I created `LocalTime` instance and called its [atDate](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/time/LocalTime.html#atDate(java.time.LocalDate)) - that method really accepts `LocalDate` parameter.

With the changes in this PR this version of the test passes OK.

**Comment by jdunkerley** [comment]
Think this would convert any Atom where first field is a `LocalDate`, which feels wider.

The original problem, I was meaning was coercing `Standard.Base.Data.Time.Date` to `java.time.LocalDate` and vice versa when we cross over the polyglot boundary.

I don't believe this fix would do anything when coming back with a LocalDate to Enso.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java
Comment (kustosz) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]: Yeah this seems too permissive. Our atoms are algebraic datatypes, `fields` is the array of all their members. So if I have `type User (created_at : Date) (name : Text) (email : Text)`, this code makes it a date now.
I see two options here:
1. We do it through an interface, i.e. this delegates to an `is_date` method (defined in Enso), `asDate` delegates to `as_date`. This is very nice because it makes things extensible, but there may be accidental clashes. Maybe something like `is_polyglot_date`, to make the Enso-side method less ambiguous in what it does.
2. We verify the constructor is one of the few we know about.

I'd lean towards 1., as I can see value in first-class interop in this language and don't want to presume "what we've built into stdlib is all anyone will ever need"

File: engine/runtime/src/test/scala/org/enso/interpreter/test/semantic/DateTest.scala
Comment (kustosz) [inline engine/runtime/src/test/scala/org/enso/interpreter/test/semantic/DateTest.scala]: Why the `Debug.eval` here?
Also, I'd rather we place these in `tests/Test` – this here is a test suite for very low-level stuff, mostly used to bootstrap the language to the point of it being able to test itself. I'd even hesitantly call it "deprecated" now.

**Review by kustosz — CHANGES_REQUESTED** [review]
We need to put more thought into API design here.

**Comment by jdunkerley** [comment]
I agree with @kustosz suggestion.

I think we probably need to make Date/DateTime/Time part of the built in space, but it would be also helpful to control how an Enso Atom is converted across the boundary. This clearly needs some API design.

File: engine/runtime/src/test/scala/org/enso/interpreter/test/semantic/DateTest.scala
Comment (JaroslavTulach) [reply engine/runtime/src/test/scala/org/enso/interpreter/test/semantic/DateTest.scala]: I admit, I just copied `EvalTest.scala` and it uses `Debug.eval` for some reason.

I was searching for a test for using `org.graalvm.polyglot.Context.eval` as that was my first understanding of what _java polyglot boundary_ may mean. Tests for `org.graalvm.polyglot` usage deserve to be written in Java, I assume. Do we have such tests somewhere? Conversion of `java.time.LocalDate` would belong there.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]: Thanks for the review. Yes delegation to `fields[0]` was just an attempt which would really go bad with the `type User (created_at : Date)....`. Thanks for the two options.

In contrary to you, I lean towards 2. Certainly as a first shot implementation. I assume it is going to be easier - also, with respect to the _opposite direction of the conversion_...

File: engine/runtime/src/test/scala/org/enso/interpreter/test/semantic/DateTest.scala
Comment (JaroslavTulach) [reply engine/runtime/src/test/scala/org/enso/interpreter/test/semantic/DateTest.scala]: I removed `Debug.eval` in [c373669](https://github.com/enso-org/enso/pull/3374/commits/c373669a197d6a212a72bcf56d4294dcf4ab6de2).

**Comment by JaroslavTulach** [comment]
> I don't believe this fix would do anything when coming back with a LocalDate to Enso.

Right, it didn't do anything. Thanks for pointing that out, James. Since [d4a8e31](https://github.com/enso-org/enso/pull/3374/commits/d4a8e31fc856be56095a62dcf9b117c5fbd09aa1) it is tested.

> I think we probably need to make Date/DateTime/Time part of the built in space, 

As soon as the engine encounters a _foreign object_ that represents a date, it needs to wrap it/convert it/mimic it by Enso `Date`. That means either that the Engine knows about the `Date` atom or its shape or ...

> helpful to control how an Enso Atom is converted across the boundary. This clearly needs some API design.

...there is an API between the Engine and the `Date` library that allows it to enhance foreign object that claim to be `isDate()`. 

I guess I try to hack something first and then we can think of proper API to do it nicer.

File: engine/runtime/src/test/scala/org/enso/interpreter/test/semantic/DateTest.scala
Comment (kustosz) [reply engine/runtime/src/test/scala/org/enso/interpreter/test/semantic/DateTest.scala]: > I admit, I just copied EvalTest.scala and it uses Debug.eval for some reason.

The reason is `EvalTest` is testing the `Debug.eval` method.

> Tests for org.graalvm.polyglot usage deserve to be written in Java, I assume. Do we have such tests somewhere?

To me, they test the functionality of Enso (i.e. "Enso is a language that can share values with other languages") – as evidenced by the fact that you haven't written a single line of non-trivial java in this test – just using it to eval Enso code and compare strings.

The best place to drop these would be [`Java_Interop_Spec`](https://github.com/enso-org/enso/blob/develop/test/Tests/src/Semantic/Java_Interop_Spec.enso).

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]: Only converting the right atom since [48bc4f1](https://github.com/enso-org/enso/pull/3374/commits/48bc4f16f8f85a14ff3b80adf92575ca5365d8cf)

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/interop/syntax/HostValueToEnsoNode.java
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/interop/syntax/HostValueToEnsoNode.java]: A chicken and egg problem, @jdunkerley & @kustosz!

The goal is to convert all incoming `java.time.LocalDate` to Enso's `Standard.Base.Data.Time.Date.Date` - but the Enso `Standard.Base.Data.Time.Date` is actually making calls into `java.time.LocalDate` methods and does need the unconverted reference!

Vicious circle showing a significant flaw...

File: engine/runtime/src/test/scala/org/enso/interpreter/test/semantic/DateTest.scala
Comment (JaroslavTulach) [reply engine/runtime/src/test/scala/org/enso/interpreter/test/semantic/DateTest.scala]: Done in [6267ba2](https://github.com/enso-org/enso/pull/3374/commits/6267ba20f4a34ca84ca97c0d8f2730ec5b376c28). Running:
```bash
$ sbt --java-home /graalvm-ce-java11-21.3.0/ buildEngineDistribution
$ ./distribution/bin/enso --run ./test/Tests/src/Semantic/Java_Interop_Spec.enso
```
should fail the test. After doing the same on the merge commit [dd15acf](https://github.com/enso-org/enso/pull/3374/commits/dd15acfadcf0d4cc95583f8b216a68679e8bb4f1) the test shall succeed.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/interop/syntax/HostValueToEnsoNode.java
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/interop/syntax/HostValueToEnsoNode.java]: @kustosz suggests to try to:
```
--- a/distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date.enso
+++ b/distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date.enso
@@ -20,7 +20,7 @@
 
          example_now = Date.now
 now : Date
-now = Date LocalDate.now
+now = LocalDate.now
 
 ## ALIAS Current Date
```
and the advice seems to work: [c57941b](https://github.com/enso-org/enso/pull/3374/commits/c57941b21d63ca9ee7369966d6d415479193bb59)

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date.enso
Comment (JaroslavTulach) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date.enso:23]: @kustosz idea: removing the wrapping here - it will be done in `InvokeMethodNode`.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date.enso
Comment (JaroslavTulach) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date.enso:63]: Another place to remove the explicit wrapping into `Date` atom.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date.enso
Comment (JaroslavTulach) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date.enso:256]: One more place. I hope I found all the ones - refactoring for Enso language missing ;-) CCing @jdunkerley

File: engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java]: An example for core Enso language vs. standard library interop. I hope @hubertp will be able to handle even this case in some cleaner way as part of his work on #3363

File: engine/runtime/src/main/java/org/enso/interpreter/node/callable/resolver/HostMethodCallNode.java
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/node/callable/resolver/HostMethodCallNode.java:97]: We don't want `LocalDateTime` (returns `true` for both `isDate` as well as `isTime`) to be converted to `Date` (at least not as part of this PR).

File: test/Tests/src/Semantic/Java_Interop_Spec.enso
Comment (JaroslavTulach) [inline test/Tests/src/Semantic/Java_Interop_Spec.enso:54]: Running
```bash
enso$ sbt --java-home ~/bin/graalvm-ce-java11-21.3.0/ buildEngineDistribution
...
enso$ built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/bin/enso \
  --run test/Tests/src/Semantic/Java_Interop_Spec.enso 
/Java_Interop_Spec.enso
....
Java/Enso Date:
    - Java date has Enso properties
    - send Enso date into Java
```
finishes without problems and that's what we wanted, I believe.

**Review by JaroslavTulach — COMMENTED** [review]
I guess we are ready for another round of review. Thanks in advance for your comments.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date.enso
Comment (kustosz) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date.enso:23]: Yes, that would be best

File: engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java
Comment (kustosz) [inline engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java]: I think this would be more elegant as a separate node, not a bare function

File: engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java
Comment (kustosz) [reply engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java]: Looks like he will!

File: engine/runtime/src/main/java/org/enso/interpreter/node/callable/resolver/HostMethodCallNode.java
Comment (kustosz) [reply engine/runtime/src/main/java/org/enso/interpreter/node/callable/resolver/HostMethodCallNode.java:97]: Surely wa want cases for `Time` and `Duration` here as well?

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java
Comment (kustosz) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]: Use `@CachedLibrary`

**Review by kustosz — COMMENTED** [review]
Very nice. Some nitpicks / discussion points inline. Also please make sure it doesn't make things slower.

File: engine/runtime/src/main/java/org/enso/interpreter/node/callable/resolver/HostMethodCallNode.java
Comment (jdunkerley) [reply engine/runtime/src/main/java/org/enso/interpreter/node/callable/resolver/HostMethodCallNode.java:97]: was assuming we would add these using this PR as guide. Seemed like something we could take of Jaroslav's list

Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/node/callable/resolver/HostMethodCallNode.java:97]: I'd be glad if I could merge this PR, finally. Letting @jdunkerley , @hubertp , co. convert the `Time` and `Duration` would help me focus on investigating the language server protocol.

File: engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java]: OK, done in: 94221c200

File: engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java]: This is wrong! We cannot cache `Atom` neither `AtomConstructor` in a `Node`, right @kustosz? I'll rewrite the code and put the cache into `Context`...

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]: Done as e13400118e212e91db8e07bbdda2ad90a1984d15

File: engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java]: Rewritten in e13400118e212e91db8e07bbdda2ad90a1984d15 to store the `AtomConstructor` in `Context`.

**Comment by JaroslavTulach** [comment]
Squashing the commits into:
- the test https://github.com/enso-org/enso/pull/3374/commits/413179eb4bc2e1194c1d29f8a5902c345f2f01ff
- the fix https://github.com/enso-org/enso/pull/3374/commits/eee9bf2fe21f333431ad81be6ba8ebe7be388c16
- additional doc/import polishing commits

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/Context.java
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/Context.java]: refactor the constant to a separate variable?

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/Context.java
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/Context.java:481]: why not `dateModule.ifPresent(module -> ...)`. is that even possible that it is not loaded if `getDateConstructor` is invoked?

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java:211]: `dateConstructor` can lead to NPE if the module is not present?

Comment (hubertp) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java:211]: also `dateConstructor.ifPreset(...).orElse(false)`

**Review by hubertp — APPROVED** [review]
lgtm

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java:211]: Since [1252ec0](https://github.com/enso-org/enso/pull/3374/commits/1252ec0c8da7bfb2af8a21cee124510824663398) `getDateConstructor()` always returns non-`null` value, but it can be `!isPresent()`.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/Context.java
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/Context.java:481]: It could happen if the Enso `Language` is manipulated from `org.graalvm.polyglot.Context` API from Java - find a `Value` representing Enso function and `value.execute(LocalDate.now())` - then the conversion can call `getDateConstructor()` without `Standard.Base.Data.Time.Date` being loaded. However I don't know the details of Enso module loading yet.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/Context.java
Comment (kustosz) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/Context.java]: I'd rather use the strongly-typed API of `Module` to get at this atom constructor, rather than the stringly-typed interop library with typechecks. `module.getScope().getLocalConstructor("Date");` replaces the whole try-catch here.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java
Comment (kustosz) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]: This looks inconsistent with google-java-format, which is the requirement for this project.

**Review by kustosz — APPROVED** [review]
Looks good, there's just one stylistic choice I disagree with and it seems like code formatting is off in general. Approving, as fixing these shouldn't require another review from me :)

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]: Related to #3394 - once there is a way to run the proper formatter and the gate fails - the proper formatting becomes a requirement. I'll spend some time trying to _guess the right fix_, but if I fail, we can probably wait for #3394 to turn the check  on and fix all the problems in a batch.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/Context.java
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/Context.java]: Done in [1252ec0](https://github.com/enso-org/enso/pull/3374/commits/1252ec0c8da7bfb2af8a21cee124510824663398), but not by using `getLocalConstructor`. That method returns `associatedType` and that's not the right constructor. Possibly related to the _mess of symbols_ as recently discussed also with @ekmett .

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/Context.java
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/Context.java]: Done in [1252ec0](https://github.com/enso-org/enso/pull/3374/commits/1252ec0c8da7bfb2af8a21cee124510824663398)

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java]: No more boundary - running at full speed at the end.

**Pull request merged**
